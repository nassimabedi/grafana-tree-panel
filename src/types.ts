export enum TreeLevelOrderMode {
  Asc = 'asc',
  Desc = 'desc',
}

export enum TreeFileldTemplateEngine {
  Simple = 'simple',
  Handlebars = 'handlebars',
}

export interface TreeOptions {
  rootName: string;
  treeFieldTemplateEngine: TreeFileldTemplateEngine;
  treeFields: string;
  serieColumn: string;
  expandLevel: number;
  showItemCount: boolean;
  orderLevels: TreeLevelOrderMode;
  enableConsoleLog?: boolean;
}

export declare type IntOrString = number | string;
