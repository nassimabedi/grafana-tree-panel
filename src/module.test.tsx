import React /*, { ReactElement }*/ from 'react';
import { configure, HTMLAttributes, mount, ReactWrapper } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import fs from 'fs';
import jsonata from 'jsonata';
import {
  FieldConfigSource,
  PanelData,
  PanelProps,
  TimeRange,
  LoadingState,
  EventBus,
  ScopedVars,
  AbsoluteTimeRange,
  FieldType,
  Field,
  ArrayVector,
} from '@grafana/data';
import { TreeLevelOrderMode, TreeOptions } from './types';
import { TreePanel, Props, idSep } from './TreePanel';

configure({ adapter: new Adapter() });

const panelPropsDefault: PanelProps = {
  id: 1,
  timeRange: ({} as unknown) as TimeRange,
  timeZone: '',
  transparent: false,
  renderCounter: 0,
  title: 'TreePanel',
  data: ({} as unknown) as PanelData,
  options: {},
  width: 800,
  height: 600,
  fieldConfig: ({} as unknown) as FieldConfigSource,
  eventBus: ({} as unknown) as EventBus,
  onOptionsChange: function (options: any): void {},
  onFieldConfigChange: function (config: FieldConfigSource<any>): void {},
  replaceVariables: function (value: string, scopedVars?: ScopedVars, format?: string | Function): string {
    return value;
  },
  onChangeTimeRange: function (timeRange: AbsoluteTimeRange): void {},
};

const optionsDefault = {
  treeFields: 'MUST_BE_SET',
  rootName: 'Pods',
  serieVariable: 'serieVariable',
  showItemCount: true,
  orderLevels: TreeLevelOrderMode.Asc,
  expandLevel: 100,
  enableConsoleLog: true,
} as TreeOptions;

const podJsonata =
  '$map(items, function($v) {{"namespace": $v.metadata.namespace, "name": $v.metadata.name, "appName": $v.metadata.labels."app.kubernetes.io/name" ? $v.metadata.labels."app.kubernetes.io/name" : ($v.metadata.labels."app" ? $v.metadata.labels."app" : "-"), "statusPhase": $v.status.phase, "containerCount": $count($v.spec.containers), "containerImage": $join($v.spec.containers[*].image, " "), "containerState": $v.status.containerStatuses[*].state ? $string($v.status.containerStatuses[*].state) : "-", "cluster": $serieVariable}})';

describe('Data processing and panel rendering', () => {
  it('simple data', () => {
    const options = {
      ...optionsDefault,
      treeFields: '${statusPhase}\n${namespace}\n${appName} ${name}\n${containerImage} ${containerState}',
    } as TreeOptions;

    let seriesData: Array<Field<string>> = [];
    const len = loadSerie(seriesData, 'test/mongo.json', 'mongo', options.serieVariable, podJsonata);

    const data = {
      series: [
        {
          name: 'mongo.json',
          fields: seriesData,
          length: len,
        },
      ],
      timeRange: ({} as unknown) as TimeRange,
      state: LoadingState.Done,
    } as PanelData;

    const params: Props = {
      ...panelPropsDefault,
      options: options,
      data: data,
    };

    const wrapper = mount(<TreePanel {...params} />);

    const nodes = wrapper.find('.customTreeItem');
    const nodesRendered = stringNodes(nodes, false);
    console.log('stringNodes():\n' + nodesRendered);

    expect(nodesRendered).toMatchSnapshot();
  });

  it('2 containers', () => {
    const options = {
      ...optionsDefault,
      treeFields: '${namespace}\n${appName}\n(${statusPhase}) ${name} [${containerImage}]\n${containerState}',
    } as TreeOptions;

    let seriesData: Array<Field<string>> = [];
    const len = loadSerie(
      seriesData,
      'test/longhorn-system.json',
      'longhorn-system',
      options.serieVariable,
      podJsonata
    );

    const data = {
      series: [
        {
          name: 'longhorn-system.json',
          fields: seriesData,
          length: len,
        },
      ],
      timeRange: ({} as unknown) as TimeRange,
      state: LoadingState.Done,
    } as PanelData;

    const params: Props = {
      ...panelPropsDefault,
      options: options,
      data: data,
    };

    const wrapper = mount(<TreePanel {...params} />);

    const nodes = wrapper.find('.customTreeItem');
    const nodesRendered = stringNodes(nodes, false);
    console.log('stringNodes():\n' + nodesRendered);

    expect(nodesRendered).toMatchSnapshot();
  });

  it('2 series', () => {
    const options = {
      ...optionsDefault,
      treeFields: '${statusPhase}\n${namespace}\n${appName} ${name}\n${containerImage}',
    } as TreeOptions;

    let seriesDataA: Array<Field<string>> = [];
    const lenA = loadSerie(seriesDataA, 'test/redis.json', 'redis', options.serieVariable, podJsonata);

    let seriesDataB: Array<Field<string>> = [];
    const lenB = loadSerie(
      seriesDataB,
      'test/rabbitmq-system.json',
      'rabbitmq-system',
      options.serieVariable,
      podJsonata
    );

    const data = {
      series: [
        {
          name: 'redis.json',
          fields: seriesDataA,
          length: lenA,
        },
        {
          name: 'rabbitmq-system.json',
          fields: seriesDataB,
          length: lenB,
        },
      ],
      timeRange: ({} as unknown) as TimeRange,
      state: LoadingState.Done,
    } as PanelData;

    const params: Props = {
      ...panelPropsDefault,
      options: options,
      data: data,
    };

    const wrapper = mount(<TreePanel {...params} />);

    const nodes = wrapper.find('.customTreeItem');
    const nodesRendered = stringNodes(nodes, false);
    console.log('stringNodes():\n' + nodesRendered);

    expect(nodesRendered).toMatchSnapshot();
  });
});

function stringNodes(root: ReactWrapper<HTMLAttributes, any, React.Component<{}, {}, any>>, showId: boolean): string {
  let kv = new Map<string, string>();
  root.forEach((node) => {
    const value = node.prop('label');
    if (node.key() === undefined || node.key() === null || value === undefined) {
      return;
    }
    kv.set(node.key(), value);
  });

  let lines: string[] = [];
  let keys = [...kv.keys()];
  keys.sort();
  keys.forEach((key) => {
    const indent = ' '.repeat(key.split(idSep).length - 1);
    if (showId) {
      lines.push(indent + key);
    }
    lines.push(indent + kv.get(key));
  });

  return lines.join('\n');
}

function makeField(name: string, values: string[]): Field<string> {
  return {
    name: name,
    type: FieldType.string,
    values: new ArrayVector(values),
    config: {},
  };
}

function loadSerie(
  seriesData: Array<Field<string>>,
  path: string,
  name: string,
  serieVariable: string,
  jsonAta: string
): number {
  const records = jsonata(jsonAta).evaluate(JSON.parse(String(fs.readFileSync(path))), {
    serieVariable: name,
  });
  console.log(records);

  let fieldValues = new Map<string, string[]>();
  let len = 0;
  records.forEach((record: Object) => {
    len++;
    Object.entries(record).forEach(([key, value]) => {
      if (!fieldValues.has(key)) {
        fieldValues.set(key, []);
      }
      fieldValues.get(key)?.push(String(value));
    });
  });

  fieldValues.forEach((values, field) => {
    expect(values.length).toBe(len);
    seriesData.push(makeField(field, values));
  });

  return len;
}

/*
 * used only for reverse engineering
 *

function findRootItem(
  wrapper: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>
): ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>> | undefined {
  const props = wrapper.props();
  console.log(props);
  const nodeId = wrapper.prop('nodeId');
  if (nodeId !== undefined) {
    return wrapper;
  }

  let root: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>> | undefined = undefined;
  wrapper.children().forEach((child) => {
    if (root === undefined) {
      const w = findRootItem(child);
      if (w !== undefined) {
        root = w;
      }
    }
  });

  return root;
}

const stringRTree = function (item: ReactWrapper, indent: string, showId: boolean): string {
  let key = '';
  let label = '';
  let children: any[] = [];
  const props = typeof item.props === 'function' ? item.props() : item.props;
  Object.entries(props).forEach((pair) => {
    const [k, value] = pair;
    switch (k) {
      case 'key':
        key = String(value);
        break;
      case 'label':
        label = String(value);
        break;
      case 'children':
        children = value as any[];
        break;
    }
  });
  if (key === '') {
    key = typeof item.key === 'function' ? item.key() : String(item.key);
  }

  let stringItem = (showId ? indent + key + '\n' : '') + indent + label + '\n';
  if (children !== undefined) {
    children.forEach((child) => {
      const childRE = child as ReactWrapper;
      if (childRE !== undefined) {
        stringItem += stringRTree(childRE, indent + '  ', showId);
      }
    });
  }

  return stringItem;
};
*/
