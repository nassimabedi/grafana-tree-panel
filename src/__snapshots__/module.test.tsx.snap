// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data processing and panel rendering Rendering 2 containers 1`] = `
"Pods (29)
  longhorn-system.json; longhorn-system (29)
    - (9)
       (Running) engine-image-ei-fa2dfbf0-2cql6 [longhornio/longhorn-engine:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:55Z\\"}}
       (Running) engine-image-ei-fa2dfbf0-8z9kj [longhornio/longhorn-engine:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:54Z\\"}}
       (Running) engine-image-ei-fa2dfbf0-lvkvx [longhornio/longhorn-engine:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:03Z\\"}}
       (Running) instance-manager-e-7c715a5a [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:52Z\\"}}
       (Running) instance-manager-e-e691695a [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:55Z\\"}}
       (Running) instance-manager-e-ee426bee [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:33Z\\"}}
       (Running) instance-manager-r-05f75ac7 [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:33Z\\"}}
       (Running) instance-manager-r-74c1e529 [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:45Z\\"}}
       (Running) instance-manager-r-e6aec4c8 [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:52Z\\"}}
    csi-attacher (3)
       (Running) csi-attacher-5f46994f7-28t4p [longhornio/csi-attacher:v3.2.1] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:02Z\\"}}
       (Running) csi-attacher-5f46994f7-5pc5s [longhornio/csi-attacher:v3.2.1] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:45Z\\"}}
       (Running) csi-attacher-5f46994f7-cnnr9 [longhornio/csi-attacher:v3.2.1] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:53Z\\"}}
    csi-provisioner (3)
       (Running) csi-provisioner-6ccbfbf86f-qhjwm [longhornio/csi-provisioner:v2.1.2] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:02Z\\"}}
       (Running) csi-provisioner-6ccbfbf86f-rn889 [longhornio/csi-provisioner:v2.1.2] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:52Z\\"}}
       (Running) csi-provisioner-6ccbfbf86f-zjx9h [longhornio/csi-provisioner:v2.1.2] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:58Z\\"}}
    csi-resizer (3)
       (Running) csi-resizer-6dd8bd4c97-2k62f [longhornio/csi-resizer:v1.2.0] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:48Z\\"}}
       (Running) csi-resizer-6dd8bd4c97-l2rwd [longhornio/csi-resizer:v1.2.0] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:03Z\\"}}
       (Running) csi-resizer-6dd8bd4c97-qdx52 [longhornio/csi-resizer:v1.2.0] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:49Z\\"}}
    csi-snapshotter (3)
       (Running) csi-snapshotter-86f65d8bc-56nd7 [longhornio/csi-snapshotter:v3.0.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:01Z\\"}}
       (Running) csi-snapshotter-86f65d8bc-qn6p8 [longhornio/csi-snapshotter:v3.0.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:58Z\\"}}
       (Running) csi-snapshotter-86f65d8bc-wrn5k [longhornio/csi-snapshotter:v3.0.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:54Z\\"}}
    longhorn-csi-plugin (3)
        (Running) longhorn-csi-plugin-hzlbr [longhornio/csi-node-driver-registrar:v2.3.0 longhornio/longhorn-manager:v1.2.3] (1)
                  [{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:20:00Z\\"}},{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:57Z\\"}}]
        (Running) longhorn-csi-plugin-npxkc [longhornio/csi-node-driver-registrar:v2.3.0 longhornio/longhorn-manager:v1.2.3] (1)
                  [{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:34Z\\"}},{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:22Z\\"}}]
        (Running) longhorn-csi-plugin-nw2lv [longhornio/csi-node-driver-registrar:v2.3.0 longhornio/longhorn-manager:v1.2.3] (1)
                  [{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:23Z\\"}},{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:16:52Z\\"}}]
    longhorn-driver-deployer (1)
       (Running) longhorn-driver-deployer-784546d78d-nz9qv [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:19Z\\"}}
    longhorn-manager (3)
       (Running) longhorn-manager-5cl94 [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:01Z\\"}}
       (Running) longhorn-manager-9jgkt [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:21Z\\"}}
       (Running) longhorn-manager-z4cr5 [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:53Z\\"}}
    longhorn-ui (1)
       (Running) longhorn-ui-9fdb94f9-w84gf [longhornio/longhorn-ui:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:18:02Z\\"}}"
`;

exports[`Data processing and panel rendering Rendering 2 containers, handlebars 1`] = `
"Pods (29)
  longhorn-system.json; longhorn-system (29)
    - (9)
       (Running) engine-image-ei-fa2dfbf0-2cql6 [longhornio/longhorn-engine:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:55Z\\"}}
       (Running) engine-image-ei-fa2dfbf0-8z9kj [longhornio/longhorn-engine:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:54Z\\"}}
       (Running) engine-image-ei-fa2dfbf0-lvkvx [longhornio/longhorn-engine:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:03Z\\"}}
       (Running) instance-manager-e-7c715a5a [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:52Z\\"}}
       (Running) instance-manager-e-e691695a [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:55Z\\"}}
       (Running) instance-manager-e-ee426bee [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:33Z\\"}}
       (Running) instance-manager-r-05f75ac7 [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:33Z\\"}}
       (Running) instance-manager-r-74c1e529 [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:45Z\\"}}
       (Running) instance-manager-r-e6aec4c8 [longhornio/longhorn-instance-manager:v1_20211210] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:52Z\\"}}
    csi-attacher (3)
       (Running) csi-attacher-5f46994f7-28t4p [longhornio/csi-attacher:v3.2.1] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:02Z\\"}}
       (Running) csi-attacher-5f46994f7-5pc5s [longhornio/csi-attacher:v3.2.1] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:45Z\\"}}
       (Running) csi-attacher-5f46994f7-cnnr9 [longhornio/csi-attacher:v3.2.1] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:53Z\\"}}
    csi-provisioner (3)
       (Running) csi-provisioner-6ccbfbf86f-qhjwm [longhornio/csi-provisioner:v2.1.2] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:02Z\\"}}
       (Running) csi-provisioner-6ccbfbf86f-rn889 [longhornio/csi-provisioner:v2.1.2] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:52Z\\"}}
       (Running) csi-provisioner-6ccbfbf86f-zjx9h [longhornio/csi-provisioner:v2.1.2] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:58Z\\"}}
    csi-resizer (3)
       (Running) csi-resizer-6dd8bd4c97-2k62f [longhornio/csi-resizer:v1.2.0] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:48Z\\"}}
       (Running) csi-resizer-6dd8bd4c97-l2rwd [longhornio/csi-resizer:v1.2.0] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:03Z\\"}}
       (Running) csi-resizer-6dd8bd4c97-qdx52 [longhornio/csi-resizer:v1.2.0] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:49Z\\"}}
    csi-snapshotter (3)
       (Running) csi-snapshotter-86f65d8bc-56nd7 [longhornio/csi-snapshotter:v3.0.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:01Z\\"}}
       (Running) csi-snapshotter-86f65d8bc-qn6p8 [longhornio/csi-snapshotter:v3.0.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:58Z\\"}}
       (Running) csi-snapshotter-86f65d8bc-wrn5k [longhornio/csi-snapshotter:v3.0.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:54Z\\"}}
    longhorn-csi-plugin (3)
        (Running) longhorn-csi-plugin-hzlbr [longhornio/csi-node-driver-registrar:v2.3.0 longhornio/longhorn-manager:v1.2.3] (1)
                  [{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:20:00Z\\"}},{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:57Z\\"}}]
        (Running) longhorn-csi-plugin-npxkc [longhornio/csi-node-driver-registrar:v2.3.0 longhornio/longhorn-manager:v1.2.3] (1)
                  [{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:34Z\\"}},{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:22Z\\"}}]
        (Running) longhorn-csi-plugin-nw2lv [longhornio/csi-node-driver-registrar:v2.3.0 longhornio/longhorn-manager:v1.2.3] (1)
                  [{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:23Z\\"}},{\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:16:52Z\\"}}]
    longhorn-driver-deployer (1)
       (Running) longhorn-driver-deployer-784546d78d-nz9qv [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:19Z\\"}}
    longhorn-manager (3)
       (Running) longhorn-manager-5cl94 [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:19:01Z\\"}}
       (Running) longhorn-manager-9jgkt [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:21Z\\"}}
       (Running) longhorn-manager-z4cr5 [longhornio/longhorn-manager:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:53Z\\"}}
    longhorn-ui (1)
       (Running) longhorn-ui-9fdb94f9-w84gf [longhornio/longhorn-ui:v1.2.3] (1)
             {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:18:02Z\\"}}"
`;

exports[`Data processing and panel rendering Rendering 2 series 1`] = `
"Pods (9)
  Pending (1)
    rabbitmq-system.json; rabbitmq-system (1)
      messaging-topology-operator messaging-topology-operator-f9c69d45b-gxmj6 (1)
         rabbitmqoperator/messaging-topology-operator:1.2.1
  Running (8)
    rabbitmq-system.json; rabbitmq-system (3)
      rabbitmq-cluster-operator rabbitmq-cluster-operator-7cbf865f89-qwwwj (1)
         rabbitmqoperator/cluster-operator:1.10.0
      rabbitmq rabbitmq-server-0 (1)
         rabbitmq:3.8.26-management
      rabbitmq rabbitmq-server-1 (1)
         rabbitmq:3.8.26-management
    redis.json; redis (5)
      redisoperator redisoperator-56d6888cc-ks84t (1)
         quay.io/spotahome/redis-operator:latest
      vcc rfr-vcc-0 (1)
         redis:6.2.6-alpine
      vcc rfr-vcc-1 (1)
         redis:6.2.6-alpine
      vcc rfs-vcc-5cc6bf796c-g9mnr (1)
         redis:6.2.6-alpine
      vcc rfs-vcc-5cc6bf796c-mmrkr (1)
         redis:6.2.6-alpine"
`;

exports[`Data processing and panel rendering Rendering 2 series, handlebars 1`] = `
"Pods (9)
  Pending (1)
    rabbitmq-system.json; rabbitmq-system (1)
      messaging-topology-operator messaging-topology-operator-f9c69d45b-gxmj6 (1)
         rabbitmqoperator/messaging-topology-operator:1.2.1
  Running (8)
    rabbitmq-system.json; rabbitmq-system (3)
      rabbitmq-cluster-operator rabbitmq-cluster-operator-7cbf865f89-qwwwj (1)
         rabbitmqoperator/cluster-operator:1.10.0
      rabbitmq rabbitmq-server-0 (1)
         rabbitmq:3.8.26-management
      rabbitmq rabbitmq-server-1 (1)
         rabbitmq:3.8.26-management
    redis.json; redis (5)
      redisoperator redisoperator-56d6888cc-ks84t (1)
         quay.io/spotahome/redis-operator:latest
      vcc rfr-vcc-0 (1)
         redis:6.2.6-alpine
      vcc rfr-vcc-1 (1)
         redis:6.2.6-alpine
      vcc rfs-vcc-5cc6bf796c-g9mnr (1)
         redis:6.2.6-alpine
      vcc rfs-vcc-5cc6bf796c-mmrkr (1)
         redis:6.2.6-alpine"
`;

exports[`Data processing and panel rendering Rendering simple data 1`] = `
"Pods (5)
  Failed (1)
    mongo.json; mongo (1)
      prometheus-mongodb-exporter mongodb-exporter-prometheus-mongodb-exporter-test-connection (1)
                   busybox {\\"terminated\\":{\\"containerID\\":\\"cri-o://5a762a4d71092cc2cb4e059240997ca747fdfcecc5ddbc9568a2a2b1a1d4985e\\",\\"exitCode\\":1,\\"finishedAt\\":\\"2021-12-17T08:52:19Z\\",\\"reason\\":\\"Error\\",\\"startedAt\\":\\"2021-12-17T08:52:18Z\\"}}
  Running (4)
    mongo.json; mongo (4)
      - percona-server-mongodb-operator-fcc5c8d6-sqb8m (1)
             percona/percona-server-mongodb-operator:1.10.0 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:18Z\\"}}
      percona-server-mongodb vcc-rs0-0 (1)
             percona/percona-server-mongodb:4.4.10-11 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:22:07Z\\"}}
      percona-server-mongodb vcc-rs0-1 (1)
             percona/percona-server-mongodb:4.4.10-11 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:22:43Z\\"}}
      prometheus-mongodb-exporter mongodb-exporter-prometheus-mongodb-exporter-6dcdd8c8fc-zqffr (1)
             ssheehy/mongodb-exporter:0.10.0 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:55Z\\"}}"
`;

exports[`Data processing and panel rendering Rendering simple data, handlebars 1`] = `
"Pods (5)
  Failed (1)
    mongo.json; mongo (1)
      prometheus-mongodb-exporter mongodb-exporter-prometheus-mongodb-exporter-test-connection (1)
                   busybox {\\"terminated\\":{\\"containerID\\":\\"cri-o://5a762a4d71092cc2cb4e059240997ca747fdfcecc5ddbc9568a2a2b1a1d4985e\\",\\"exitCode\\":1,\\"finishedAt\\":\\"2021-12-17T08:52:19Z\\",\\"reason\\":\\"Error\\",\\"startedAt\\":\\"2021-12-17T08:52:18Z\\"}}
  Running (4)
    mongo.json; mongo (4)
      - percona-server-mongodb-operator-fcc5c8d6-sqb8m (1)
             percona/percona-server-mongodb-operator:1.10.0 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:18Z\\"}}
      percona-server-mongodb vcc-rs0-0 (1)
             percona/percona-server-mongodb:4.4.10-11 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:22:07Z\\"}}
      percona-server-mongodb vcc-rs0-1 (1)
             percona/percona-server-mongodb:4.4.10-11 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:22:43Z\\"}}
      prometheus-mongodb-exporter mongodb-exporter-prometheus-mongodb-exporter-6dcdd8c8fc-zqffr (1)
             ssheehy/mongodb-exporter:0.10.0 {\\"running\\":{\\"startedAt\\":\\"2022-01-20T17:17:55Z\\"}}"
`;

exports[`Data processing and panel rendering Rendering simple data, handlebars, PodStatus 1`] = `
"Pods (5)
  Error (1)
    mongo.json; mongo (1)
      prometheus-mongodb-exporter mongodb-exporter-prometheus-mongodb-exporter-test-connection (1)
         busybox Terminated: exit(1)
  Running (4)
    mongo.json; mongo (4)
      - percona-server-mongodb-operator-fcc5c8d6-sqb8m (1)
         percona/percona-server-mongodb-operator:1.10.0 <none>
      percona-server-mongodb vcc-rs0-0 (1)
         percona/percona-server-mongodb:4.4.10-11 <none>
      percona-server-mongodb vcc-rs0-1 (1)
         percona/percona-server-mongodb:4.4.10-11 <none>
      prometheus-mongodb-exporter mongodb-exporter-prometheus-mongodb-exporter-6dcdd8c8fc-zqffr (1)
         ssheehy/mongodb-exporter:0.10.0 <none>"
`;
